### PR TITLE
New test runner implementation

### PR DESCRIPTION
**Pull request by rojer**
Simplified, added support for serializing state and resuming (not used yet but will be).

### CODE REVIEW

**Comment by rojer** [comment]
> will **decrease** coverage by `1.66%`.

new runner is 91% covered by tests but since more code is deleted than added, overall coverage goes down.

File: pkg/runner/test_runner.go
Comment (rihter007) [inline pkg/runner/test_runner.go:358]: Because we acquire the lock in the goroutine above, we get blocked here until that goroutine finishes (swch channel is closed)

Comment (rojer) [reply pkg/runner/test_runner.go:358]: not quite. cond.Wait() releases the lock while waiting.

Comment (rihter007) [reply pkg/runner/test_runner.go:358]: Sorry, it seems to be ok because of tr.cond.Wait() will release the mutex

**Comment by rojer** [comment]
clarified the description a bit, on feedback from @rihter007

**Comment by marcoguerri** [comment]
I haven't started full review yet, it will take a while. What this is missing is emissions of `TargetIn` (https://github.com/facebookincubator/contest/blob/master/pkg/runner/test_runner_route.go#L88) and `TargetOut` events (https://github.com/facebookincubator/contest/blob/master/pkg/runner/test_runner_route.go#L204), which are then used to rebuild the status of the job (https://github.com/facebookincubator/contest/blob/master/pkg/runner/job_status.go#L23-L28) from any instance, even those that do not own the job.

This is duplication of how we keep track of the state and we should converge into using the same approach both for resume and and Status queries.

**Comment by rojer** [comment]
@marcoguerri thanks! indeed, i did not pay sufficient attention to the events that should be emitted. this should be fixed now - i added TargetIn, TargetOut, TargetErr and made sure a TestError is emitted in the right cases.

File: pkg/runner/test_runner.go
Comment (tfg13) [inline pkg/runner/test_runner.go:377]:
```suggestion
		// Early check for pause or cancelation.
```

File: pkg/runner/test_runner.go
Comment (tfg13) [inline pkg/runner/test_runner.go:625]:
```suggestion
				break loop
```

File: pkg/runner/test_runner.go
Comment (tfg13) [inline pkg/runner/test_runner.go:690]:
```suggestion
	// if the pipeline was resumed).
```

**Review by tfg13 — COMMENTED** [review]
Mostly nits. Since it is a super large change, it would be good if someone else would also read it as well.

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go]: actually, on further reflection - no, we don't want to break out of the loop just yet. as the comment above says, we want to make sure error channel also gets processed, which is exactly why we replace outCh with a new one and wait for ss.errCh to be closed as well.

File: pkg/runner/goroutine_leak_check.go
Comment (rojer) [reply pkg/runner/goroutine_leak_check.go:37]: done

File: pkg/jobmanager/jobmanager.go
Comment (rojer) [reply pkg/jobmanager/jobmanager.go:160]: gone

File: pkg/runner/test_runner.go
Comment (tfg13) [reply pkg/runner/test_runner.go]: Ah wasn't sure what you want to break out of, I guess you meant the select then?
Would `continue loop` be clearer in that case?

Comment (rojer) [reply pkg/runner/test_runner.go]: great idea, indeed that would make the intent more clear. done.

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:440]: What is `recd`?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:41]: mves -> moves

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:688]: I see you lock the mutex for the whole runMonitor execution. How `targetRunners` works in this case?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:303]: Could you rename this variable to something more meaningful, please?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:296]: `ctx` is not used.

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:229]: Could you please elaborate in this comment?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:250]: Better to use structured logs. See `WithField`/`WithFields`. It allows to perform smarter log-based analytics afterwards.

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:521]: Why this recover is required? Could you leave a comment, please?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:532]: Out of curiosity, why we do not initialize `test.TestStepChannels` right inside `stepState` (and instead we define three separate channels)?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:625]: `loop` is not required here

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:634]: It looks racy for me:
What will happen if `outCh` is closed, but `errCh` contains an error? Or this is intended?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:623]: Why not `outCh = nil`?

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:645]: It does not look like proper draining: I guess there could be multiple events in `ss.errCh`.

File: pkg/runner/test_runner.go
Comment (xaionaro) [inline pkg/runner/test_runner.go:651]: So we timed-out but not interrupted the loop? Is this correct?

**Review by xaionaro — APPROVED** [review]

**Comment by xaionaro** [comment]
Just a warning: the quality of my review was poor this time.

File: pkg/cerrors/cerrors.go
Comment (rojer) [reply pkg/cerrors/cerrors.go:69]: done

File: pkg/runner/job_runner.go
Comment (rojer) [reply pkg/runner/job_runner.go:190]: done

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:30]: done

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:625]: it's not but i think it's more readable that way

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:623]: indeed. i forgot that reading from nil channel blocks.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:645]: out and err are unbuffered and are closed only after step's run function has exited or if we gave up on waiting for it to exit.
step is not allowed to leave its run function without successfully processing targets so we don't need to care about some targets getting stuck. if the step left its run method, it means all targets have been processed.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:651]: we are interrupting the loop - just below this we forcibly close the output channel even though the run method is still running precisely to release the reader. the reader will close error channel and exit, and it will also release the goroutine we started above. as a result, the only leak will be of the goroutine stuck in the step's Run method (this is verified in the test using CheckLeakedGoRoutines)

**Comment by xaionaro** [comment]
P.S.: Also mutex-logic looks dangerous in multiple ways. I guess it might cause a deadlock somewhere in future (when somebody will modify the code).

File: pkg/runner/goroutine_leak_check.go
Comment (rojer) [reply pkg/runner/goroutine_leak_check.go:52]: added a test

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:634]: that's precisely the expected sequence: out is closed, then error. when out is closed, we continue to wait for error to be closed, as explained in the comment above.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:95]: ok, done

**Comment by rojer** [comment]
> P.S.: Also mutex-logic looks dangerous in multiple ways. I guess it might cause a deadlock somewhere in future (when somebody will modify the code).

well, this is overly general. previous implementation didn't use mutexes but was a maze of channels that was impossible to understand.
in this refactoring i tried to keep locking simple. in particular by having only one mutex, even though using separate per-target and per-step locks might've been more efficient but would make code considerably more complicated.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:191]: This new `targetState` doesn't get backported in `tr.targets`?

EDIT: nevermind, it happens down below.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:217]: Does `waitStepRunners` guarantee that all steps have returned? If so, `stepCancel` should not be necessary. If not, steps should be awaited for if a cancellation signal is being propagated. We should keep track of any step that does not return within the timeout and flag it accordingly.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:239]: If there is even only one target that failed in any step, we are declaring the test not resumable (but continuing the loop after we declare `resumeOk` false)? It should be fine to fail a target in a step, and still be able to resume the test.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:281]: Ho do we get here? `runMonitor` has returned, `waitStepRunners` has returned, so I assume all steps have  returned (without an error? otherwise `runErr` would have been not nil and we would have returned before?) How do we get into this state, where we consider steps terminated, monitor terminated, but still targets might be in-flight?

**Review by marcoguerri — COMMENTED** [review]
Apologies for the late review. In fact, I am catching up only now, and I am also far from being done (for my reference, I shall resume from `waitStepRunners`). Even if this is pushed, I still want to finish leaving a first pass of comments today.

**Comment by marcoguerri** [comment]
> but was a maze of channels that was impossible to understand.

There was a lot of room for simplification in the previous implementation, in particular:
* Not using context made several things overly complicated
* All results, were waited  by the test runner. There was a lot of room for simplification to give distribute responsibility to routing blocks, and simplify test runner (gave it a shot in https://github.com/facebookincubator/contest/pull/122/commits/c53e4269ef940b51fd6eed7d85dbf5656fe84000)
* Running pipelines back to back (test + cleanup) also added quite a bit of complexity, that could be removed (commit above does this as well). 

This just to say that in my opinion this solution has its own traps, as it makes data flow implicit (via mutex and global structure) rather than explicit, via channels. No questioning that the previous solution could be simplified, of course, but I personally find this solution not significantly easier to follow, with respect to the previous version with some unnecessary complexity removed.

**Comment by xaionaro** [comment]
> This just to say that in my opinion this solution has its own traps, as it makes data flow implicit (via mutex and global structure) rather than explicit, via channels. No questioning that the previous solution could be simplified, of course, but I personally find this solution not significantly easier to follow, with respect to the previous version with some unnecessary complexity removed.

I have exactly the same impression.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:217]: you are right, this is not needed as by this point all the step runners have either exited or we gave up waiting on them to exit.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:239]: stepState.runErr aka stepErr is not about target errors, it is

// Runner error, returned from Run() or an error condition detected by the reader.

i.e. it's an error condition that aborts entire run. presence of such errors making run non-resumable makes sense.

File: pkg/runner/test_runner.go
Comment (rojer) [reply pkg/runner/test_runner.go:281]: that's if plugin "swallowed" a target - it returned from Run with no error but never reported result for a target that was injected.

**Comment by rojer** [comment]
@marcoguerri sorry, i should have waited for your approval before merging. but i created https://github.com/facebookincubator/contest/pull/212 so we can continue the review there.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [reply pkg/runner/test_runner.go:239]: Well, ok, not sure why I missed that here we were referring to an error of the whole step.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:603]: Looks like `safeCloseOutCh` and `safeCloseErrCh` can be merged together and accept an argument from outside.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:640]: If `outCh` indicates that the test step is guaranteed to have returned, then whoever signaled completion by closing `outCh` should have closed also `errCh`. Responsibility of closing these channels should not be spread across multiple entities.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:662]: This should not be necessary. It should be `stepRunner` guaranteeing to us that step termination is protected against `shutdownTimeout`. We shouldn't be doing it in multiple places.

File: pkg/runner/test_runner.go
Comment (marcoguerri) [inline pkg/runner/test_runner.go:731]: This truly deserves a comment. I don't understand this check, why we break down below, and why `if !ok`, we just wait for condition variable to be signaled and continue looping into the following step.

**Comment by marcoguerri** [comment]
Tried to finish review of at least `test_runner.go`. Looks like I could not comment in #212 outside of the scope of that specific PR (which is really not ideal).

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:286]: done

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:298]: this has to do with how waiting on `sync.Condition` works. when you want to wait for some internal state protected by a mutex to change (in this case `stepRunning` and `readerRunning`), you acquire a mutex, test the state and if it is not to your satisfaction, you invoke `tr.cond.Wait()` which releases the mutex and then suspends until signaled, at which point it re-locks and returns, letting you re-examine the state again. we signal `tr.cond` when things change (runner exits, rader exits of target handler exits).

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:350]: the step runner context will eventually be canceled (via defer stepCancel())

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:381]: fixed

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:376]: you are right, fixed

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:731]: the comment above explains the purpose of the entire loop - check if all the targets have either made it past the injection phase or terminated.
the first condition checks if the target has terminated or not (if terminated, there will not be a result channel).
this one checks if it made past the target injection phase.
it's perfectly fine for this to not be the case, in this case we just wait for things to change (for condition to be signaled).

i simplified the check a bit, since we break out on first !ok, there's no reason to test for it (a leftover from previous implementation where we'd iterate all targets anyway).

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:603]: yes, except channels have different types and i couldn't find a way to cast them to something common.

File: pkg/runner/test_runner.go
Comment (rojer9-fb) [reply pkg/runner/test_runner.go:640]: this will be simplified when we have one channel for results (as discussed). i'll leave it as is for now.

**Pull request merged**
